File: yogi-dotnet/test/TimerTest.cs
using System;
using System.Net;
using Xunit;

namespace test
{
    public class TimerTest : TestCase
    {
        Yogi.Context context = new Yogi.Context();

        [Fact]
        public void Start()
        {
            Yogi.Timer timer = new Yogi.Timer(context);

            bool called = false;
            timer.Start(TimeSpan.FromMilliseconds(1), (res) => {
                Assert.IsType<Yogi.Success>(res);
                Assert.Equal(Yogi.ErrorCode.Ok, res.ErrorCode);
                called = true;
            });

            GC.Collect();

            while (!called)
            {
                context.RunOne();
            }

            Assert.True(called);

            GC.KeepAlive(timer);
        }

        [Fact]
        public void Cancel()
        {
            Yogi.Timer timer = new Yogi.Timer(context);

            Assert.False(timer.Cancel());

            bool called = false;
            timer.Start(Yogi.InfiniteTimeSpan, (res) => {
                Assert.IsType<Yogi.Failure>(res);
                Assert.Equal(Yogi.ErrorCode.Canceled, res.ErrorCode);
                called = true;
            });

            GC.Collect();
            Assert.True(timer.Cancel());
            GC.Collect();

            while (!called)
            {
                context.RunOne();
            }

            Assert.True(called);

            GC.KeepAlive(timer);
        }
    }
}

File: yogi-cpp/test/public_api_tests/timer_test.cc
#include "../common.h"

#include <yogi_core.h>

class TimerTest : public ::testing::Test {
 protected:
  yogi::ContextPtr context_ = yogi::Context::Create();
  yogi::TimerPtr timer_ = yogi::Timer::Create(context_);
};

TEST_F(TimerTest, Start) {
  bool called = false;
  timer_->Start(yogi::Duration::FromMilliseconds(1), [&](auto& res) {
    EXPECT_NO_THROW(dynamic_cast<const yogi::Success&>(res));
    EXPECT_EQ(res.GetErrorCode(), yogi::ErrorCode::kOk);
    called = true;
  });

  while (!called) {
    context_->RunOne();
  }

  EXPECT_TRUE(called);
}

TEST_F(TimerTest, Cancel) {
  EXPECT_FALSE(timer_->Cancel());

  bool called = false;
  timer_->Start(yogi::Duration::kInfinity, [&](auto& res) {
    EXPECT_NO_THROW(dynamic_cast<const yogi::Failure&>(res));
    EXPECT_EQ(res.GetErrorCode(), yogi::ErrorCode::kCanceled);
    called = true;
  });

  EXPECT_TRUE(timer_->Cancel());

  while (!called) {
    context_->RunOne();
  }

  EXPECT_TRUE(called);
}
